lib/truffle/java/java.rb

```diff
@@ -0,0 +1,188 @@
+    end
+
+    alias_method :eql?, :==
+
+    def equal?(another)
+      nil != @value and JavaObject === another and @value.equals(another.value)
+    end
+
+    def java_type
+      @value.getClass().getName()
+    end
+
+    def java_class
+      JavaClass.get(@value.getClass())
+    end
+
+    def length
+      raise TypeError, "Not a java array"
+    end
+
+    def java_proxy?
+      true
+    end
+
+    def synchronized(&block)
+      return @value.__synchronize__(block) if @value != nil
+      NULL_LOCK.synchronize block
+    end
+
+    def marshal_dump
+    end
+
+    def marshal_load
+    end
+  end
+
+  class JavaArray < JavaObject
+    def length
+      java.lang.reflect.Array.getLength(@value)
+    end
+  end
+
+  class JavaClass
+  end
+
+  class JavaPackage < Module
+
+    attr_reader :package_name
+    attr_reader :children
+
+    def self.new(*name_parts)
+      const_name = name_parts.reduce("") { |memo, obj| memo + obj.capitalize }
+      return ::Java.const_get(const_name) if ::Java.const_defined?(const_name, false)
+      super(*name_parts)
+    end
+
+    def initialize(*name_parts)
+      ruby_name = name_parts.reduce("") do |memo, obj| memo + obj.capitalize end
+      java_name = name_parts.join(".")
+      @package_name = java_name
+      @children = {}
+      ::Java.const_set(ruby_name, self)
+      parent = JavaPackage.new(*name_parts.first(name_parts.size - 1)) unless name_parts.size <= 1
+      parent.add_child(name_parts.last, self) unless parent == nil
+      super()
+    end
+
+    def self.capitalised_name(name)
+      name = ""
+      package_name.split(".").each do |s|
+        name += s.capitalize
+      end
+    end
+
+    def to_s
+      @package_name
+    end
+
+    def inspect
+      super.to_s
+    end
+
+    def const_missing(name)
+       method_missing(name)
+    end
+
+    def add_child(package, name)
+      @children[name] = package
+    end
+
+    def method_missing(name, *args)
+      if args.size != 0
+        raise ArgumentError, "#{self} does not have a method #{name} with #{args.size} arguments."
+      end
+
+      val = @children[name]
+
+      return val if val != nil
+
+      val = JavaUtilities.get_relative_package_or_class(self, name)
+      @children[name] = val
+
+      if name[0] == name[0].upcase && !val.kind_of?(JavaPackage)
+        const_set(name, val)
+      end
+      val
+    end
+  end
+
+  class JavaProxy
+  end
+
+  @packages = {}
+
+  def self.const_missing name
+    JavaPackage.new(name.to_s)
+  end
+
+  def self.method_missing(name, *args)
+    if args.size != 0
+      raise ArgumentError, "Java does not have a method #{name} with #{args.size} arguments."
+    end
+
+    val = @packages[name.capitalize]
+
+    return val if val != nil
+
+    val = JavaUtilities.get_package_module_dot_format(name.to_s)
+
+    @packages[name.capitalize] = val
+  end
+
+  def self.java_to_ruby obj
+    # We'll do something here, not sure what yet.
```

Comment:
Leave a initial and dated TODO, so we know who to ask what we're going to do and how long since we thought to do it.